### DISCUSSION TITLE
Consider not encoding the `class` attribute for the `attr()` filter

### DISCUSSION THREAD

**Discussion by mmikkel** [craftcms/cms#12886] [Ideas]
Currently, the `|attr` filter [encodes all attribute values](https://github.com/yiisoft/yii2/blob/0792736b3512776a70bb96e47050855f55edf063/framework/helpers/BaseHtml.php#L2006) - which is generally not a problem and expected/wanted behavior.

But, I think it might be a good idea to make an exception for the `class` attribute.

A specific example: I recently ran into a gotcha using [Tailwind's arbitrary variants feature](https://tailwindcss.com/docs/hover-focus-and-other-states#using-arbitrary-variants). Consider this markup, sitting in a template partial `button.twig`:

```html
<button type="button" class="[&[disabled]]:opacity-50">Click me</button>
```

That `[&[disabled]]:opacity-50` class name certainly looks weird! But, it's perfectly valid HTML, and a valid Tailwind utility 🤷 (the class will reduce the opacity for the button to 50%, if it has a `disabled` attribute).

However, if this markup is rendered via the `attr` filter, the ampersand will be encoded and the class will no longer do what it's intended to do … because Tailwind doesn't recognize that encoded ampersand as the `&` placeholder.

I.e. this Twig:

```twig
{{ include('button.twig')|attr({ disabled: true }) }}
```
will render this HTML:

```html
<button type="button" class="[&amp;[disabled]]:opacity-50" disabled>Click me</button>
```

I'm sure there are other examples of technically valid class names being encoded and thus potentially breaking.

Since the attribute encoding happens in [yii\helpers\Html::renderTagAttributes()](https://github.com/yiisoft/yii2/blob/0792736b3512776a70bb96e47050855f55edf063/framework/helpers/BaseHtml.php#L1996), I don't know if a fix in Craft's `|attr` filter is even feasible. But, especially when considering Tailwind's popularity, I expect this is something that a lot of people is going to run into going forward so I figured it was worth pointing out.

**Comment by mmikkel** [comment]
Doh. I just realized that the `attr()` filter encoding the `class` attribute isn't actually a problem 🙃 

The real gotcha is that the `attr()` filter seems to be double-encoding existing attributes. In my case the `<button>` was originally created using the `{% tag %}` tag, which means that its attributes were already encoded (including the `class`). And when the `attr()` filter is applied, that double-encodes the ampersand, so the actual HTML looks like this: `<button type="button" class="[&amp;amp;[disabled]]:opacity-50" disabled>Click me</button>` - which is why things break.

This looks more like a bug than expected behavior to me, so I created issue #12887 for this instead.

**Comment by brandonkelly** [comment]
This is now fixed in 3.8.2 and 4.4.2 (via a77f32c66d66d066fc1a1261c76b9af4f0db0a80).
